pkg/kv/kvserver/loqrecovery/apply.go

```diff
@@ -330,14 +338,101 @@ func CommitReplicaChanges(batches map[roachpb.StoreID]storage.Batch) (ApplyUpdat
 		}
 	}
-	if failed {
+	if len(updateErrors) > 0 {
 		return report, errors.Errorf(
 			"failed to commit update to one or more stores: %s", strings.Join(updateErrors, "; "))
 	}
 	return report, nil
 }
+
+// MaybeApplyPendingRecoveryPlan applies loss of quorum recovery plan if it is
+// staged in planStore. Changes would be applied to engines when their
+// identities match storeIDs of replicas in the plan.
+// Plan applications errors like mismatch of store with plan, inability to
+// deserialize values ets are only reported to logs and application status but
+// are not propagated to caller. Only serious errors that imply misconfiguration
+// or planStorage issues are propagated.
+// Regardless of application success or failure, staged plan would be removed.
+func MaybeApplyPendingRecoveryPlan(
+	ctx context.Context, planStore PlanStore, engines []storage.Engine, clock timeutil.TimeSource,
+) error {
+	if len(engines) < 1 {
+		return nil
+	}
+
+	applyPlan := func(nodeID roachpb.NodeID, plan loqrecoverypb.ReplicaUpdatePlan) error {
+		log.Infof(ctx, "applying staged loss of quorum recovery plan %s", plan.PlanID)
+		batches := make(map[roachpb.StoreID]storage.Batch)
+		for _, e := range engines {
+			ident, err := kvstorage.ReadStoreIdent(ctx, e)
+			if err != nil {
+				return errors.Wrap(err, "failed to read store ident when trying to apply loss of quorum recovery plan")
+			}
+			b := e.NewBatch()
+			defer b.Close()
+			batches[ident.StoreID] = b
+		}
+		prepRep, err := PrepareUpdateReplicas(ctx, plan, uuid.DefaultGenerator, clock.Now(), nodeID, batches)
+		if err != nil {
+			return err
+		}
+		if len(prepRep.MissingStores) > 0 {
+			log.Warningf(ctx, "loss of quorum recovery plan application expected stores on the node %s",
+				strutil.JoinIDs("s", prepRep.MissingStores))
+		}
+		_, err = CommitReplicaChanges(batches)
+		if err != nil {
+			// This is not very good as are in a partial success situation, but we don't
+			// have a good solution other than report that as error. Let the user
+			// decide what to do next.
+			return err
+		}
+		return nil
+	}
+
+	plan, exists, err := planStore.LoadPlan()
+	if err != nil {
+		// This is fatal error, we don't write application report since we didn't
+		// check the store yet.
+		return errors.Wrap(err, "failed to check if loss of quorum recovery plan is staged")
+	}
+	if !exists {
+		return nil
+	}
+
+	// First read node parameters from the first store.
+	storeIdent, err := kvstorage.ReadStoreIdent(ctx, engines[0])
+	if err != nil {
+		if errors.Is(err, &kvstorage.NotBootstrappedError{}) {
+			// This is wrong, we must not have staged plans in a non-bootstrapped
+			// node. But we can't write an error here as store init might refuse to
+			// work if there are already some keys in store.
+			log.Errorf(ctx, "node is not bootstrapped but it already has a recovery plan staged: %s", err)
+			return nil
+		}
+		return err
+	}
+
+	if err := planStore.RemovePlan(); err != nil {
```

Comment:
Should we remove this before it's successfully applied? Is it better to retry on the next restart?

Reply:
We already reported it as failed though. If we have a replica that updated after we created plan then we'll end up with plan that would fail forever not being able to apply. I think it is safer to drop it. And run another recovery if needed.